lib/ansible/modules/database/postgresql/postgresql_query.py

```diff
@@ -0,0 +1,357 @@
+      more information on the modes.
+    - Default of C(prefer) matches libpq default.
+    type: str
+    default: prefer
+    choices: [ allow, disable, prefer, require, verify-ca, verify-full ]
+  ssl_rootcert:
+    description:
+    - Specifies the name of a file containing SSL certificate authority (CA)
+      certificate(s).
+    - If the file exists, the server's certificate will be
+      verified to be signed by one of these authorities.
+    type: str
+notes:
+- The default authentication assumes that you are either logging in as or
+  sudo'ing to the postgres account on the host.
+- To avoid "Peer authentication failed for user postgres" error,
+  use postgres user as a I(become_user).
+- This module uses psycopg2, a Python PostgreSQL database adapter. You must
+  ensure that psycopg2 is installed on the host before using this module. If
+  the remote host is the PostgreSQL server (which is the default case), then
+  PostgreSQL must also be installed on the remote host. For Ubuntu-based
+  systems, install the postgresql, libpq-dev, and python-psycopg2 packages
+  on the remote host before using this module.
+requirements: [ psycopg2 ]
+author:
+- Felix Archambault (@archf)
+- Andrew Klychkov (@Andersson007)
+- Will Rouesnel (@wrouesnel)
+'''
+
+EXAMPLES = r'''
+- name: Simple select query to acme db
+  postgresql_query:
+    db: acme
+    query: SELECT version()
+
+- name: Select query to db acme with positional arguments and non-default credentials
+  postgresql_query:
+    db: acme
+    login_user: django
+    login_password: mysecretpass
+    query: SELECT * FROM acme WHERE id = %s AND story = %s
+    positional_args:
+    - 1
+    - test
+
+- name: Select query to test_db with named_args
+  postgresql_query:
+    db: test_db
+    query: SELECT * FROM test WHERE id = %(id_val)s AND story = %(story_val)s
+    named_args:
+      id_val: 1
+      story_val: test
+
+- name: Insert query to db test_db
+  postgresql_query:
+    db: test_db
+    query: INSERT INTO test_db (id, story) VALUES (2, 'my_long_story')
+
+- name: Run queries from SQL script
+  postgresql_query:
+    db: test_db
+    path_to_script: /var/lib/pgsql/test.sql
+    positional_args:
+    - 1
+'''
+
+RETURN = r'''
+query:
+    description: Query that was tried to be executed.
+    returned: always
+    type: str
+    sample: 'SELECT * FROM bar'
+statusmessage:
+    description: Attribute containing the message returned by the command.
+    returned: always
+    type: str
+    sample: 'INSERT 0 1'
+query_result:
+    description:
+    - List of dictionaries in column:value form representing returned rows.
+    returned: changed
+    type: list
+    sample: [{"Column": "Value1"},{"Column": "Value2"}]
+rowcount:
+    description: Number of affected rows.
+    returned: changed
+    type: int
+    sample: 5
+'''
+
+import os
+
+try:
+    import psycopg2
+    HAS_PSYCOPG2 = True
+except ImportError:
+    HAS_PSYCOPG2 = False
```

Comment:
Where are you checking this?

In Ansible 2.8 we added https://github.com/ansible/community/issues/346#issuecomment-461037623 which hopefully will give give clearer error messages to users.

Reply:
Good notice! Thank you, fixed

Reply:
@gundalow , is it necessary to use traceback here ?
```
import traceback

PSYCOPG2_IMP_ERR = None
...
    PSYCOPG2_IMP_ERR = traceback.format_exc()
```
I'm asking because I've recently heard that it's not necessary.
If we use module.fail_json(msg=missing_required_lib('psycopg2'), exception=PSYCOPG2_IMP_ERR)
the output is a little different:
```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_postgresql_ping_payload_HnnFhj/__main__.py", line 113, in <module>
    import psycopg2
ImportError: No module named psycopg2

spblnx176 | FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "db": null, 
            "login_host": "", 
            "login_password": "", 
            "login_unix_socket": "", 
            "login_user": "postgres", 
            "port": 5432, 
            "ssl_mode": "prefer", 
            "ssl_rootcert": null
        }
    }, 
    "msg": "Failed to import the required Python library (psycopg2) on spblnx176.lan's Python /usr/bin/python. Please read module documentation and install in the appropriate location"
}
```
module.fail_json(msg=missing_required_lib('psycopg2') without traceback
```
The full traceback is:
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_postgresql_ping_payload_y3izFs/__main__.py", line 111, in <module>
    import psycopg2

spblnx176 | FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "db": null, 
            "login_host": "", 
            "login_password": "", 
            "login_unix_socket": "", 
            "login_user": "postgres", 
            "port": 5432, 
            "ssl_mode": "prefer", 
            "ssl_rootcert": null
        }
    }, 
    "msg": "Failed to import the required Python library (psycopg2) on spblnx176.lan's Python /usr/bin/python. Please read module documentation and install in the appropriate location"
}
```
Should I add traceback to all my new modules including this or not ? (now I use just module.fail_json(msg=missing_required_lib('psycopg2') here)

Reply:
I think what you've got it good, thanks